### PR TITLE
Gossip self cleanup

### PR DESCRIPTION
**Pull request by art-gor**
Removing non actual peer information from gossip instances.
This self-cleanup is needed because some nodes use ephemeral peer ids

### CODE REVIEW

File: include/libp2p/peer/address_repository/inmem_address_repository.hpp [hunk 1]
```diff
@@ -59,13 +59,16 @@ namespace libp2p::peer {
     std::unordered_set<PeerId> getPeers() const override;
 
    private:
-    bool isNewDnsAddr(const multi::Multiaddress &ma);
-
     using ttlmap = std::unordered_map<multi::Multiaddress, Clock::time_point>;
     using ttlmap_ptr = std::shared_ptr<ttlmap>;
+    using peer_db = std::unordered_map<PeerId, ttlmap_ptr>;
```
Comment (igor-egorov) [inline include/libp2p/peer/address_repository/inmem_address_repository.hpp:64]: oof, double map :)

File: include/libp2p/protocol/gossip/gossip.hpp [hunk 2]
```diff
@@ -48,13 +49,25 @@ namespace libp2p::protocol::gossip {
...
+    std::chrono::milliseconds heartbeat_interval_msec{1000};
+
+    /// Ban interval between dial attempts to peer
+    std::chrono::milliseconds ban_interval_msec{60000};
+
+    /// Max number of dial attempts before peer is forgotten
+    unsigned max_dial_attempts = 10;
+
+    /// Expiration of gossip peers' addresses in address repository
+    std::chrono::milliseconds address_expiration_msec{3600000};
```
Comment (igor-egorov) [inline include/libp2p/protocol/gossip/gossip.hpp:70]: I would suggest using compile-time expressions and comments to values for more clarity.
```suggestion
    std::chrono::milliseconds address_expiration_msec{60 * 60 * 1000}; // 1 hour
```

or even better
```suggestion
    std::chrono::milliseconds address_expiration_msec{std::chrono::hours(1)};
```

File: include/libp2p/protocol/gossip/gossip.hpp [hunk 3]
```diff
@@ -48,13 +49,25 @@ namespace libp2p::protocol::gossip {
...
-    unsigned seen_cache_lifetime_msec = 60000;
+    /// Topic's message seen cache lifetime
+    std::chrono::milliseconds seen_cache_lifetime_msec{60000};
 
-    unsigned heartbeat_interval_msec = 1000;
+    /// Heartbeat interval
+    std::chrono::milliseconds heartbeat_interval_msec{1000};
+
+    /// Ban interval between dial attempts to peer
+    std::chrono::milliseconds ban_interval_msec{60000};
```
Comment (igor-egorov) [inline include/libp2p/protocol/gossip/gossip.hpp:64]: The same idea as for the value of 3600000 ms
```suggestion
    std::chrono::milliseconds ban_interval_msec{std::chrono::minutes(1)};
```

File: src/protocol/gossip/impl/connectivity.cpp [hunk 4]
```diff
@@ -216,44 +204,27 @@ namespace libp2p::protocol::gossip {
...
-      unban(ctx);
-    } else {
-      connectable_peers_.erase(ctx->peer_id);
-    }
+    connectable_peers_.erase(ctx->peer_id);
 
     peer::PeerInfo pi = host_->getPeerRepository().getPeerInfo(ctx->peer_id);
-    if (ctx->dial_to) {
-      pi.addresses = {ctx->dial_to.value()};
-    }
```
Comment (igor-egorov) [inline src/protocol/gossip/impl/connectivity.cpp:243]: Hmm, cannot understand why this code is gone.

File: src/protocol/gossip/impl/connectivity.cpp [hunk 5]
```diff
@@ -403,28 +445,32 @@ namespace libp2p::protocol::gossip {
...
+          [this](const PeerContextPtr &ctx) { flush(ctx); });
     }
-    // clang-format on
 
     writable_peers_low_latency_.clear();
     writable_peers_on_heartbeat_.clear();
+
+    if (ts >= addresses_renewal_time_) {
+      addresses_renewal_time_ = scheduler_->now()
+          + config_.address_expiration_msec - std::chrono::milliseconds(500);
```
Comment (igor-egorov) [inline src/protocol/gossip/impl/connectivity.cpp:467]: Not clear why we are subtracting 500ms here. Probably this should be a named constant with a comment.

**Review by igor-egorov — APPROVED** [review]

File: src/protocol/gossip/impl/connectivity.cpp (hunk 4 above)
Comment (art-gor) [reply src/protocol/gossip/impl/connectivity.cpp:243]: this field is gone, address repo is used now

**Review by igor-egorov — APPROVED** [review]

File: include/libp2p/protocol/gossip/gossip.hpp [hunk 6]
```diff
@@ -48,13 +49,30 @@ namespace libp2p::protocol::gossip {
     bool echo_forward_mode = false;
 
     /// Read or write timeout per whole network operation
-    unsigned rw_timeout_msec = 10000;
+    std::chrono::milliseconds rw_timeout_msec{std::chrono::seconds(10)};
```
Comment (xDimon) [inline include/libp2p/protocol/gossip/gossip.hpp:52]: Suggest to remove `_msec` suffix

File: src/peer/address_repository/inmem_address_repository.cpp [hunk 7]
```diff
@@ -63,11 +63,20 @@ namespace libp2p::peer {
     return false;
   }
 
+  InmemAddressRepository::peer_db::iterator
+  InmemAddressRepository::findOrInsert(const PeerId &p) {
+    auto it = db_.find(p);
+    if (it == db_.end()) {
+      it = db_.emplace(p, std::make_shared<ttlmap>()).first;
+    }
+    return it;
```
Comment (xDimon) [inline src/peer/address_repository/inmem_address_repository.cpp:72]:
```suggestion
    return db_.emplace(p, std::make_shared<ttlmap>()).first;
```

File: src/protocol/gossip/impl/gossip_core.cpp [hunk 8]
```diff
@@ -213,7 +204,7 @@ namespace libp2p::protocol::gossip {
       from->message_builder->addMessage(*msg_found.value(), msg_id);
       connectivity_->peerIsWritable(from, true);
     } else {
-      log_.warn("wanted message not in cache");
+      log_.debug("wanted message not in cache");
```
Comment (xDimon) [inline src/protocol/gossip/impl/gossip_core.cpp:207]:
```suggestion
      SL_DEBUG(log_, "wanted message not in cache");
```

File: src/protocol/gossip/impl/stream.cpp [hunk 9]
```diff
@@ -175,14 +172,15 @@ namespace libp2p::protocol::gossip {
     );
     // clang-format on
 
-    if (timeout_ > 0) {
-      timeout_handle_ =
-          scheduler_.schedule(timeout_, [self_wptr = weak_from_this(), this] {
+    if (timeout_ > std::chrono::milliseconds::zero()) {
```
Comment (xDimon) [inline src/protocol/gossip/impl/stream.cpp:175]: ?
```suggestion
    if (timeout_ > Time::zero()) {
```

**Review by xDimon — APPROVED** [review]

**Pull request merged**